# Inspect Data
______

In [135]:
!pip install -U pandas numpy chardet
import pandas as pd
import numpy as np
import re

# Connect to Google Drive
from google.colab import drive
drive.mount('/content/drive')

# Load the data
df = pd.read_csv("/content/drive/MyDrive/cosmetic_p.csv", encoding="utf-8-sig")

# Check the first five rows
display(df.head())

# Inspect the types of products
display(df['Label'].value_counts())

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Label           brand                                           name  \
0  Moisturizer          LA MER                                Crème de la Mer   
1  Moisturizer           SK-II                       Facial Treatment Essence   
2  Moisturizer  DRUNK ELEPHANT                     Protini™ Polypeptide Cream   
3  Moisturizer          LA MER                    The Moisturizing Soft Cream   
4  Moisturizer    IT COSMETICS  Your Skin But Better™ CC+™ Cream with SPF 50+   

   price  rank                                        ingredients  \
0    175   4.1  Algae (Seaweed) Extract, Mineral Oil, Petrolat...   
1    179   4.1  Galactomyces Ferment Filtrate (Pitera), Butyle...   
2     68   4.4  Water, Dicaprylyl Carbonate, Glycerin, Ceteary...   
3    175   3.8  Algae (Seaweed) Extract, Cyclopentasiloxane, P...   
4     38   4.1  Water, Snail Secretion Filtrate, Phenyl Trimet...   

   Combination  Dry  Normal  Oily  Sensitive  
0            1    1       1     1          1  
1            1    1       1     1          1  
2            1    1       1     1          0  
3            1    1       1     1          1  
4            1    1       1     1          1

Label
Moisturizer    298
Cleanser       281
Face Mask      266
Treatment      248
Eye cream      209
Sun protect    170
Name: count, dtype: int64

In [136]:
df.shape

(1472, 11)

In [137]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1472 entries, 0 to 1471
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Label        1472 non-null   object 
 1   brand        1472 non-null   object 
 2   name         1472 non-null   object 
 3   price        1472 non-null   int64  
 4   rank         1472 non-null   float64
 5   ingredients  1472 non-null   object 
 6   Combination  1472 non-null   int64  
 7   Dry          1472 non-null   int64  
 8   Normal       1472 non-null   int64  
 9   Oily         1472 non-null   int64  
 10  Sensitive    1472 non-null   int64  
dtypes: float64(1), int64(6), object(4)
memory usage: 126.6+ KB


In [138]:
df['brand'].nunique()

116

In [139]:
df['brand'].unique()

array(['LA MER', 'SK-II', 'DRUNK ELEPHANT', 'IT COSMETICS', 'TATCHA',
       "KIEHL'S SINCE 1851", 'FRESH', 'BELIF', 'SUNDAY RILEY', 'FARMACY',
       'FIRST AID BEAUTY', 'CLINIQUE', 'BAREMINERALS', 'SHISEIDO',
       'CHARLOTTE TILBURY', 'ORIGINS', 'OLEHENRIKSEN', 'JOSIE MARAN',
       'FARSÁLI', 'LANEIGE', 'DR. JART+', 'GLOW RECIPE', 'HERBIVORE',
       'TARTE', 'CAUDALIE', 'YOUTH TO THE PEOPLE', 'BIOSSANCE',
       'ESTÉE LAUDER', 'SON & PARK', 'AMOREPACIFIC', 'BOBBI BROWN',
       'ERBORIAN', 'SMASHBOX', 'GLAMGLOW', 'PETER THOMAS ROTH',
       'JACK BLACK', 'PHILOSOPHY', 'PERRICONE MD', 'LANCÔME', 'ALGENIST',
       'MAKE UP FOR EVER', 'KORRES', 'MURAD', 'GUERLAIN', 'DIOR',
       'KORA ORGANICS', 'DR. DENNIS GROSS SKINCARE', 'REN CLEAN SKINCARE',
       'KOPARI', 'CLARINS', 'NURSE JAMIE', 'LANCER', "L'OCCITANE",
       'KATE SOMERVILLE', 'SUPERGOOP!', 'SATURDAY SKIN',
       'SEPHORA COLLECTION', 'MILK MAKEUP', 'DR. BRANDT SKINCARE',
       'J.ONE', 'KAT VON D', 'TATA HARPER', 'GO

In [140]:
df.isnull().sum()

Label          0
brand          0
name           0
price          0
rank           0
ingredients    0
Combination    0
Dry            0
Normal         0
Oily           0
Sensitive      0
dtype: int64

ไม่มี missing data




In [141]:
df.nunique()

Label             6
brand           116
name           1472
price           146
rank             29
ingredients    1337
Combination       2
Dry               2
Normal            2
Oily              2
Sensitive         2
dtype: int64

- Label = 6 ประเภทของผลิตภัณฑ์
- brand = 116 แบรนด์
- name = 1472 ชื่อสินค้า
- rank = 0-29 rank
- ingredients = 1,337 อย่าง


In [142]:
df.duplicated().sum()

np.int64(0)

ไม่มี duplicate

# Data Cleaning
_____

# Cleaning brand

In [143]:
# Cleaning brand
df['brand'] = df['brand'].str.strip()                      # ลบช่องว่าง
df['brand'] = df['brand'].str.lower()                      # แปลงเป็นตัวพิมพ์เล็ก
df['brand'] = df['brand'].apply(lambda x: re.sub(r'[^a-z0-9\s]', '', str(x)))  # ลบอักขระพิเศษ

df['brand'].value_counts()

brand
clinique              79
sephora collection    66
shiseido              63
origins               54
murad                 47
                      ..
tom ford               1
kaplan md              1
black up               1
urban decay            1
dermaflash             1
Name: count, Length: 116, dtype: int64

In [144]:
df['brand'].unique()

array(['la mer', 'skii', 'drunk elephant', 'it cosmetics', 'tatcha',
       'kiehls since 1851', 'fresh', 'belif', 'sunday riley', 'farmacy',
       'first aid beauty', 'clinique', 'bareminerals', 'shiseido',
       'charlotte tilbury', 'origins', 'olehenriksen', 'josie maran',
       'farsli', 'laneige', 'dr jart', 'glow recipe', 'herbivore',
       'tarte', 'caudalie', 'youth to the people', 'biossance',
       'este lauder', 'son  park', 'amorepacific', 'bobbi brown',
       'erborian', 'smashbox', 'glamglow', 'peter thomas roth',
       'jack black', 'philosophy', 'perricone md', 'lancme', 'algenist',
       'make up for ever', 'korres', 'murad', 'guerlain', 'dior',
       'kora organics', 'dr dennis gross skincare', 'ren clean skincare',
       'kopari', 'clarins', 'nurse jamie', 'lancer', 'loccitane',
       'kate somerville', 'supergoop', 'saturday skin',
       'sephora collection', 'milk makeup', 'dr brandt skincare', 'jone',
       'kat von d', 'tata harper', 'goto', 'boscia'

In [145]:
# ✅ 1. ตรวจว่ามีช่องว่างหน้า-หลังไหม
spaces = df[df['brand'].str.match(r'^\s+|\s+$', na=False)]

# ✅ 2. ตรวจว่ามีค่าเป็น NaN หรือค่าว่างไหม
missing = df[df['brand'].isna() | (df['brand'].astype(str).str.strip() == '')]

# ✅ 3. ตรวจตัวพิมพ์ใหญ่-เล็กไม่สม่ำเสมอ
brand_case = df['brand'].str.lower().nunique() != df['brand'].nunique()

# ✅ 4. ตรวจว่ามีอักขระพิเศษหรือไม่ปกติไหม (นอกจากตัวอักษร, ตัวเลข, space, +, & , ')
special_chars = df[df['brand'].str.contains(r"[^a-zA-Z0-9+&' ]", na=False)]

# ✅ 5. สรุปผล
print("🧾 Data Cleanliness Report for column: 'brand'\n")

print(f"1️⃣ ช่องว่างหน้า-หลัง: {'พบ' if not spaces.empty else 'ไม่พบ'}")
print(f"2️⃣ ค่าว่างหรือ NaN: {'พบ' if not missing.empty else 'ไม่พบ'}")
print(f"3️⃣ ตัวพิมพ์ใหญ่/เล็กไม่สม่ำเสมอ: {'พบ' if brand_case else 'ไม่พบ'}")
print(f"4️⃣ อักขระพิเศษ (เช่น ! , . , / , …): {'พบ' if not special_chars.empty else 'ไม่พบ'}")

print("\n📋 สรุปจำนวนแบรนด์ทั้งหมด:", df['brand'].nunique())

# (ถ้าอยากดูตัวอย่างแถวที่มีปัญหา)
if not spaces.empty:
    print("\n🔹 ตัวอย่างแบรนด์ที่มีช่องว่าง:")
    print(spaces['brand'].unique())

if not special_chars.empty:
    print("\n🔹 ตัวอย่างแบรนด์ที่มีอักขระพิเศษ:")
    print(special_chars['brand'].unique())


🧾 Data Cleanliness Report for column: 'brand'

1️⃣ ช่องว่างหน้า-หลัง: ไม่พบ
2️⃣ ค่าว่างหรือ NaN: ไม่พบ
3️⃣ ตัวพิมพ์ใหญ่/เล็กไม่สม่ำเสมอ: ไม่พบ
4️⃣ อักขระพิเศษ (เช่น ! , . , / , …): ไม่พบ

📋 สรุปจำนวนแบรนด์ทั้งหมด: 116


# Cleaning 'name'

In [146]:
df['name'] = df['name'].str.strip().str.lower()
df['name'] = df['name'].apply(lambda x: re.sub(r'\s+', ' ', str(x)))
df['name'] = df['name'].apply(lambda x: re.sub(r'[™®]', '', str(x)))


In [147]:
df['name'].value_counts()

name
crème de la mer                                                                 1
facial treatment essence                                                        1
protini polypeptide cream                                                       1
the moisturizing soft cream                                                     1
your skin but better cc+ cream with spf 50+                                     1
                                                                               ..
yoghurt nourishing fluid veil face sunscreen broad spectrum spf 30              1
daily deflector waterlight broad spectrum spf 50+ pa+++ anti-aging sunscreen    1
self tan dry oil spf 50                                                         1
pro light self tan bronzing mist                                                1
dermaprotect daily defense broad spectrum spf 50+                               1
Name: count, Length: 1472, dtype: int64

# Cleaning 'ingredients'

In [148]:
# 🔹 ลบเครื่องหมาย * ทุกแถวในคอลัมน์ ingredients ก่อนตรวจ
df['ingredients'] = df['ingredients'].astype(str).str.replace('*', '', regex=False)

In [149]:
df.loc[5, "ingredients"]

'Water, Saccharomyces/Camellia Sinensis Leaf/Cladosiphon Okamuranus/Rice Ferment Filtrate, Dimethicone, Propanediol, Glycerin, Diglycerin, Diphenylsiloxy Phenyl Trimethicone, Gold, Belamcanda Chinensis Root Extract, Rosa Multiflora Fruit Extract, Houttuynia Cordata Extract, Sophora Angustifolia Root Extract, Sodium Hyaluronate, Lecithin, Pistacia Lentiscus (Mastic) Gum, Sodium Chloride, Sodium Citrate, Mica, Peg-9 Polydimethylsiloxyethyl Dimethicone, Dimethicone/Peg-10/15 Crosspolymer, Dimethicone/Phenyl Vinyl Dimethicone Crosspolymer, Disodium Edta, Tin Oxide, Titanium Dioxide, Butylene Glycol, Ethylhexylglycerin, Parfum/Fragrance, Alcohol, Phenoxyethanol. Hadasei-3.'

In [150]:
# พิสูจน์ว่า ingredients มีคำซ้ำ
def has_duplicate_ingredients(text):
    parts = [p.strip().lower() for p in str(text).split(",")]
    return len(parts) != len(set(parts))

# ถ้ามีค่า True จำนวนมาก → แปลว่า “มี ingredients ซ้ำในแถว” จริง
df['has_dup_ing'] = df['ingredients'].apply(has_duplicate_ingredients)
df['has_dup_ing'].value_counts()


has_dup_ing
False    1414
True       58
Name: count, dtype: int64

In [151]:
# รายการแถวที่ ingredients ซ้ำ
dup_ing_df = df[df['has_dup_ing'] == True][['brand','name','ingredients']]
dup_ing_df.head(10)

brand                                               name  \
15        drunk elephant                                        the littles   
19                  skii                            r.n.a. power face cream   
46             herbivore                pink cloud rosewater moisture crème   
51   youth to the people                      adaptogen deep moisture cream   
128               korres  wild rose + vitamin c advanced brightening sle...   
187             algenist                        overnight restorative cream   
195        saturday skin                      waterfall glacier water cream   
243   dr brandt skincare  do not age with dr. brandt moisturizing neck c...   
345       drunk elephant               baby pekee bar + juju bar travel duo   
372              dr jart                                   ceramidin liquid   

                                           ingredients  
15   Beste™ No.9 Jelly Cleanser:\r\r\nWater, Sodium...  
19   Water, Glycerin, Galactomyces Ferment Filtrate...  
46   Water, Aloe Barbadensis (Aloe Vera) Leaf Juice...  
51   Purified Water, Helianthus Annuus (Sunflower) ...  
128  Water, Glycerin, Glyceryl Stearate Citrate, Di...  
187  Water, Isopropyl Isostearate, Butyrospermum Pa...  
195  Water, Glycerin, Propanediol, 1,2-Hexanediol, ...  
243  Water, Caprylic/Capric Triglyceride, Squalane,...  
345  -Heilmoor Clay: Tones and detoxifies the skin....  
372  Water, Dipropylene Glycol, Alcohol, Betaine, P...

In [152]:
# "คำที่ซ้ำ" ในแต่ละแถว
def find_duplicates(text):
    parts = [p.strip().lower() for p in str(text).split(",")]
    dups = [p for p in parts if parts.count(p) > 1]
    return list(set(dups))

df['dup_list'] = df['ingredients'].apply(find_duplicates)
df[df['has_dup_ing'] == True][['brand','name','dup_list']].head(10)

brand                                               name  \
15        drunk elephant                                        the littles   
19                  skii                            r.n.a. power face cream   
46             herbivore                pink cloud rosewater moisture crème   
51   youth to the people                      adaptogen deep moisture cream   
128               korres  wild rose + vitamin c advanced brightening sle...   
187             algenist                        overnight restorative cream   
195        saturday skin                      waterfall glacier water cream   
243   dr brandt skincare  do not age with dr. brandt moisturizing neck c...   
345       drunk elephant               baby pekee bar + juju bar travel duo   
372              dr jart                                   ceramidin liquid   

                                              dup_list  
15   [pentylene glycol, sodium hyaluronate crosspol...  
19                                  [peg-100 stearate]  
46                                  [cetearyl alcohol]  
51                                [ethylhexylglycerin]  
128                [rosmarinus officinalis (rosemary)]  
187                                      [dimethicone]  
195                                    [isohexadecane]  
243                        [calcium hydroxymethionine]  
345  [titanium dioxide (ci 77891), water, disodium ...  
372                                 [triethylhexanoin]

In [153]:
# “ล้างซ้ำ” ในขั้นตอน Data Cleaning
def clean_ingredients(text):
    txt = str(text).lower().replace("’", "'")
    parts = [p.strip(" .") for p in txt.split(",")]
    seen, result = set(), []
    for p in parts:
        if p and p not in seen:
            seen.add(p)
            result.append(p)
    return " | ".join(result)

df["ingredients_list"] = df["ingredients"].apply(clean_ingredients)

In [154]:
def has_duplicate_after_clean(text):
    parts = [p.strip() for p in str(text).split('|')]
    return len(parts) != len(set(parts))

df['dup_after_clean'] = df['ingredients_list'].apply(has_duplicate_after_clean)
df['dup_after_clean'].value_counts()

dup_after_clean
False    1472
Name: count, dtype: int64

In [155]:
df[df['has_dup_ing'] == True][['ingredients', 'ingredients_list']].head(10)

ingredients  \
15   Beste™ No.9 Jelly Cleanser:\r\r\nWater, Sodium...   
19   Water, Glycerin, Galactomyces Ferment Filtrate...   
46   Water, Aloe Barbadensis (Aloe Vera) Leaf Juice...   
51   Purified Water, Helianthus Annuus (Sunflower) ...   
128  Water, Glycerin, Glyceryl Stearate Citrate, Di...   
187  Water, Isopropyl Isostearate, Butyrospermum Pa...   
195  Water, Glycerin, Propanediol, 1,2-Hexanediol, ...   
243  Water, Caprylic/Capric Triglyceride, Squalane,...   
345  -Heilmoor Clay: Tones and detoxifies the skin....   
372  Water, Dipropylene Glycol, Alcohol, Betaine, P...   

                                      ingredients_list  
15   beste™ no.9 jelly cleanser:\r\r\nwater | sodiu...  
19   water | glycerin | galactomyces ferment filtra...  
46   water | aloe barbadensis (aloe vera) leaf juic...  
51   purified water | helianthus annuus (sunflower)...  
128  water | glycerin | glyceryl stearate citrate |...  
187  water | isopropyl isostearate | butyrospermum ...  
195  water | glycerin | propanediol | 1 | 2-hexaned...  
243  water | caprylic/capric triglyceride | squalan...  
345  -heilmoor clay: tones and detoxifies the skin....  
372  water | dipropylene glycol | alcohol | betaine...

In [156]:
def extract_ingredients_only(text: str) -> str:
    s = str(text)

    # 1) รวมบรรทัดให้เป็นบรรทัดเดียว (กัน \r \n แปลก ๆ)
    s = re.sub(r'[\r\n]+', ' ', s)

    # 2) หา keyword "ingredient / ingredients" ตามด้วย :, -, – แล้วตัดส่วนหน้าออก
    #    ใช้ finditer เพื่อเลือก "ตำแหน่งสุดท้าย" เผื่อมีคำนี้ซ้ำ
    pat = re.compile(r'ingredient[s]?\s*[:\-–]\s*', flags=re.IGNORECASE)
    matches = list(pat.finditer(s))
    if matches:
        s = s[matches[-1].end():]  # ตัดทุกอย่างก่อน "ingredients:" ออก

    # 3) ทำความสะอาดสเปซ/คอมมาเล็กน้อย
    s = re.sub(r'\s*,\s*', ', ', s)   # ช่องว่างรอบคอมมาให้มาตรฐาน
    s = re.sub(r'\s+', ' ', s).strip()

    return s

# ใช้งานกับ DataFrame ของคุณ
df['ingredients_cleaned'] = df['ingredients'].apply(extract_ingredients_only)

# (ตัวเลือก) ดูจำนวนแถวที่ถูกตัด prefix ออกจริง ๆ
num_changed = (df['ingredients'] != df['ingredients_cleaned']).sum()
print(f"ตัดคำอธิบายหน้า 'Ingredients:' ออกแล้ว {num_changed} แถว")

# (ตัวเลือก) ดูตัวอย่างก่อน-หลัง
# df.loc[row_index, ['ingredients','ingredients_cleaned']]


ตัดคำอธิบายหน้า 'Ingredients:' ออกแล้ว 464 แถว


In [157]:
df.loc[15, ['ingredients','ingredients_cleaned']]

ingredients            Beste™ No.9 Jelly Cleanser:\r\r\nWater, Sodium...
ingredients_cleaned    Beste™ No.9 Jelly Cleanser: Water, Sodium Laur...
Name: 15, dtype: object

In [158]:

def check_invalid_reason(text):
    raw = str(text)
    t = raw.strip().lower()

    # รวมบรรทัดให้เป็นรูปแบบเดียว
    t_nl = re.sub(r'\r\n?|\u2028|\u2029', '\n', t)

    # 1️⃣ เคสแบบ bullet/หัวข้อ (เช่น "-Vitamin E: Acts as ...")
    bullet_lines = re.findall(
        r'^\s*[-•·]?\s*[^:\n]{1,80}:\s*[^\n,]+',  # ชื่อหัวข้อ: ตามด้วยคำอธิบาย
        t_nl,
        flags=re.MULTILINE
    )
    if len(bullet_lines) >= 1:
        return f"⚠️ ข้อความแบบบูลเล็ต/หัวข้อ {len(bullet_lines)} บรรทัดที่มี ':' (คำอธิบาย ไม่ใช่สูตร)"

    # 2️⃣ ถ้ามี '-' ที่ต้นบรรทัด (แม้ไม่มี ':' ก็ถือว่าน่าสงสัย)
    if re.search(r'^\s*-\s*[A-Za-z]', t_nl, flags=re.MULTILINE):
        return "⚠️ มีเครื่องหมาย '-' นำหน้าบรรทัด (อาจเป็นคำอธิบาย ไม่ใช่สูตรจริง)"

    # 3️⃣ ข้อความโฆษณา/บรรยายทั่วไป
    if re.search(r'visit the|provides|acts as|known to|powerful|boutique|rich in|treat|heal', t):
        return "⚠️ มีข้อความโฆษณาหรือบรรยาย (e.g., provides / acts as / known to)"

    # 4️⃣ สัญลักษณ์การตลาด
    if re.search(r'\*|\#|%|100%', t):
        return "⚠️ มีสัญลักษณ์หรือคำทางการตลาด (*, %, 100%)"

    # 5️⃣ มี ':' มากกว่า 1 ตัว (backup)
    if t.count(':') > 1:
        return "⚠️ มีเครื่องหมาย ':' มากกว่า 1 ตัว (อาจมีหลายสูตรหรือคำอธิบายใน cell เดียว)"

    # 6️⃣ สูตร Excel / #NAME?
    if raw.startswith('=') or '#name?' in t:
        return "⚠️ ข้อความมีลักษณะสูตร Excel (#NAME?, =...)"

    # 7️⃣ สั้นผิดปกติ
    if len(t.split()) < 5:
        return "⚠️ จำนวนคำ น้อยเกินไป — อาจไม่ใช่สูตร ingredients จริง"

    # 8️⃣ ตัวอักษรควบคุมหรือ newline แปลก
    if re.search(r'[\x00-\x08\x0b-\x1f]+', raw):
        return "⚠️ มีตัวอักษรควบคุมซ่อนอยู่"

    return ""


In [159]:
# เพิ่มคอลัมน์ reason แสดงสาเหตุ
df['invalid_reason'] = df['ingredients'].apply(check_invalid_reason)

# flag ว่าน่าสงสัยไหม (True/False)
df['invalid_ing'] = df['invalid_reason'] != ""

# ดูตัวอย่างแถวที่น่าสงสัย
invalid_df = df[df['invalid_ing'] == True][['brand', 'name', 'ingredients', 'invalid_reason']]
invalid_df.head(10)


brand                                               name  \
7   drunk elephant                    virgin marula luxury facial oil   
15  drunk elephant                                        the littles   
26  drunk elephant               virgin marula luxury facial oil mini   
32    olehenriksen        sheer transformation perfecting moisturizer   
39         dr jart                         premium beauty balm spf 45   
55        shiseido                     eudermine revitalizing essence   
57     este lauder     micro essence skin activating treatment lotion   
73     este lauder       revitalizing supreme global anti-aging crème   
90         origins  high-potency night-a-mins oil-free resurfacing...   
91           tarte  amazonian clay bb tinted moisturizer broad spe...   

                                          ingredients  \
7   100% Unrefined Sclerocraya Birrea (Marula) Ker...   
15  Beste™ No.9 Jelly Cleanser:\r\r\nWater, Sodium...   
26  100% Unrefined Sclerocraya Birrea (Marula) Ker...   
32                    Visit the OLEHENRIKSEN boutique   
39  -Titanium Dioxide, Octinoxate, Zinc Oxide, and...   
55                        Visit the Shiseido boutique   
57  Micro Essence Skin Act Treat Lot Division: El ...   
73  Revitalizing Supreme+ Cell Pwr Crm Division: E...   
90                                      Essential Oil   
91  -Amazonian clay: Provides total balancing for ...   

                                       invalid_reason  
7         ⚠️ มีสัญลักษณ์หรือคำทางการตลาด (*, %, 100%)  
15  ⚠️ ข้อความแบบบูลเล็ต/หัวข้อ 7 บรรทัดที่มี ':' ...  
26        ⚠️ มีสัญลักษณ์หรือคำทางการตลาด (*, %, 100%)  
32  ⚠️ มีข้อความโฆษณาหรือบรรยาย (e.g., provides / ...  
39  ⚠️ ข้อความแบบบูลเล็ต/หัวข้อ 1 บรรทัดที่มี ':' ...  
55  ⚠️ มีข้อความโฆษณาหรือบรรยาย (e.g., provides / ...  
57  ⚠️ ข้อความแบบบูลเล็ต/หัวข้อ 1 บรรทัดที่มี ':' ...  
73  ⚠️ ข้อความแบบบูลเล็ต/หัวข้อ 1 บรรทัดที่มี ':' ...  
90  ⚠️ จำนวนคำ น้อยเกินไป — อาจไม่ใช่สูตร ingredie...  
91  ⚠️ ข้อความแบบบูลเล็ต/หัวข้อ 4 บรรทัดที่มี ':' ...

In [160]:
# จำนวนแถวที่ผิดรวมทั้งหมด
df['invalid_ing'].value_counts()

invalid_ing
False    1123
True      349
Name: count, dtype: int64

In [161]:
# 🔹 ลบแถวที่ invalid_ing == True ออกจาก DataFrame
df = df[~df['invalid_ing']]

# 🔹 รีเซ็ต index ใหม่ให้เรียงต่อเนื่อง (ไม่กระโดด)
df = df.reset_index(drop=True)

# 🔹 แสดงตารางตัวอย่างหลังล้างแถวออก
df.head(10)  # หรือจะใช้ df.tail(10) เพื่อดูท้ายตารางก็ได้


Label              brand  \
0  Moisturizer             la mer   
1  Moisturizer               skii   
2  Moisturizer     drunk elephant   
3  Moisturizer             la mer   
4  Moisturizer       it cosmetics   
5  Moisturizer             tatcha   
6  Moisturizer     drunk elephant   
7  Moisturizer  kiehls since 1851   
8  Moisturizer             la mer   
9  Moisturizer              fresh   

                                                name  price  rank  \
0                                    crème de la mer    175   4.1   
1                           facial treatment essence    179   4.1   
2                          protini polypeptide cream     68   4.4   
3                        the moisturizing soft cream    175   3.8   
4        your skin but better cc+ cream with spf 50+     38   4.1   
5                                    the water cream     68   4.2   
6                           lala retro whipped cream     60   4.2   
7                                 ultra facial cream     29   4.4   
8  little miss miracle limited-edition crème de l...    325   5.0   
9                   lotus youth preserve moisturizer     45   4.3   

                                         ingredients  Combination  Dry  \
0  Algae (Seaweed) Extract, Mineral Oil, Petrolat...            1    1   
1  Galactomyces Ferment Filtrate (Pitera), Butyle...            1    1   
2  Water, Dicaprylyl Carbonate, Glycerin, Ceteary...            1    1   
3  Algae (Seaweed) Extract, Cyclopentasiloxane, P...            1    1   
4  Water, Snail Secretion Filtrate, Phenyl Trimet...            1    1   
5  Water, Saccharomyces/Camellia Sinensis Leaf/Cl...            1    0   
6  Water, Glycerin, Caprylic/ Capric Triglyceride...            1    1   
7  Water, Glycerin, Cyclohexasiloxane, Squalane, ...            1    1   
8  Algae (Seaweed) Extract, Mineral Oil, Petrolat...            0    0   
9  Water, Glycerin, Propylene Glycol Dicaprylate/...            0    0   

   Normal  Oily  Sensitive  has_dup_ing dup_list  \
0       1     1          1        False       []   
1       1     1          1        False       []   
2       1     1          0        False       []   
3       1     1          1        False       []   
4       1     1          1        False       []   
5       1     1          1        False       []   
6       1     1          0        False       []   
7       1     1          1        False       []   
8       0     0          0        False       []   
9       0     0          0        False       []   

                                    ingredients_list  dup_after_clean  \
0  algae (seaweed) extract | mineral oil | petrol...            False   
1  galactomyces ferment filtrate (pitera) | butyl...            False   
2  water | dicaprylyl carbonate | glycerin | cete...            False   
3  algae (seaweed) extract | cyclopentasiloxane |...            False   
4  water | snail secretion filtrate | phenyl trim...            False   
5  water | saccharomyces/camellia sinensis leaf/c...            False   
6  water | glycerin | caprylic/ capric triglyceri...            False   
7  water | glycerin | cyclohexasiloxane | squalan...            False   
8  algae (seaweed) extract | mineral oil | petrol...            False   
9  water | glycerin | propylene glycol dicaprylat...            False   

                                 ingredients_cleaned invalid_reason  \
0  Algae (Seaweed) Extract, Mineral Oil, Petrolat...                  
1  Galactomyces Ferment Filtrate (Pitera), Butyle...                  
2  Water, Dicaprylyl Carbonate, Glycerin, Ceteary...                  
3  Algae (Seaweed) Extract, Cyclopentasiloxane, P...                  
4  Water, Snail Secretion Filtrate, Phenyl Trimet...                  
5  Water, Saccharomyces/Camellia Sinensis Leaf/Cl...                  
6  Water, Glycerin, Caprylic/ Capric Triglyceride...                  
7  Water, Glycerin, Cyclohexasiloxane, Squalane, ...                  
8  Algae (

In [162]:
# แถวที่ไม่ถูกลบ (คือ invalid_ing == False)
valid_df = df[df['invalid_ing'] == False]

# แสดงจำนวนแถวที่ไม่ถูกลบ
print(f"✅ จำนวนแถวที่เหลือ (ไม่ถูกลบ): {len(valid_df)} แถว")

✅ จำนวนแถวที่เหลือ (ไม่ถูกลบ): 1123 แถว


In [163]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1123 entries, 0 to 1122
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Label                1123 non-null   object 
 1   brand                1123 non-null   object 
 2   name                 1123 non-null   object 
 3   price                1123 non-null   int64  
 4   rank                 1123 non-null   float64
 5   ingredients          1123 non-null   object 
 6   Combination          1123 non-null   int64  
 7   Dry                  1123 non-null   int64  
 8   Normal               1123 non-null   int64  
 9   Oily                 1123 non-null   int64  
 10  Sensitive            1123 non-null   int64  
 11  has_dup_ing          1123 non-null   bool   
 12  dup_list             1123 non-null   object 
 13  ingredients_list     1123 non-null   object 
 14  dup_after_clean      1123 non-null   bool   
 15  ingredients_cleaned  1123 non-null   o

# Cleaning 'skin_type' (target)

“dry”, “oily”, “combination”, “sensitive”

In [164]:
# จากปัญหาที่เจอคือบางแถว “dry”, “oily”, “combination”, “sensitive” เป็น 0 ทั้งหมด
# ทำให้สามารถนำไปวิเคราะห์ mapping กับสภาพผิวได้ จึงต้องตัดแถวนั้นๆออก
# สร้าง mask สำหรับแถวที่ทุกค่าเป็น 0
mask = (
    (df['Combination'] == 0) &
    (df['Dry'] == 0) &
    (df['Normal'] == 0) &
    (df['Oily'] == 0) &
    (df['Sensitive'] == 0)
)

# เก็บแถวที่ตรงเงื่อนไข (จะถูกลบ)
deleted_rows = df[mask]

# แสดงจำนวนและข้อมูลของแถวที่ถูกลบ
print(f"🧹 จำนวนแถวที่ถูกลบ: {len(deleted_rows)} แถว\n")
print("📋 แถวที่ถูกลบ:")
print(deleted_rows)

# ลบแถวออกจาก DataFrame
df = df[~mask]

# แสดงจำนวนแถวสุดท้ายที่เหลือ
print(f"\n✅ จำนวนแถวที่เหลือหลังลบแล้ว: {len(df)} แถว")


🧹 จำนวนแถวที่ถูกลบ: 312 แถว

📋 แถวที่ถูกลบ:
            Label                         brand  \
8     Moisturizer                        la mer   
9     Moisturizer                         fresh   
16    Moisturizer                         fresh   
19    Moisturizer                  bareminerals   
20    Moisturizer                      shiseido   
...           ...                           ...   
1116  Sun protect                 vita liberata   
1118  Sun protect               kate somerville   
1120  Sun protect               kate somerville   
1121  Sun protect                 vita liberata   
1122  Sun protect  st tropez tanning essentials   

                                                   name  price  rank  \
8     little miss miracle limited-edition crème de l...    325   5.0   
9                      lotus youth preserve moisturizer     45   4.3   
16                      rose deep hydration moisturizer     40   4.4   
19    complexion rescue tinted moisturizer broad spe...

# Normalize price

เป้าหมาย: เก็บไว้ในรูปแบบตัวเลขเพื่อใช้คำนวณหรือแนะนำสินค้าในงบประมาณ

In [165]:
# 2️⃣ แปลงราคา USD → THB (ใช้ค่าแลกเปลี่ยนประมาณ 36.5)
usd_to_thb = 36.5
df['price'] = df['price'].astype(float) * usd_to_thb

print("✅ Normalized file created successfully! (name fixed + price converted)")


✅ Normalized file created successfully! (name fixed + price converted)


In [166]:
df['price'] = df['price'].apply(lambda x: re.sub(r'[^\d.]', '', str(x)))
df['price'] = pd.to_numeric(df['price'], errors='coerce')

df['price'].head()


0    6387.5
1    6533.5
2    2482.0
3    6387.5
4    1387.0
Name: price, dtype: float64

# 6️⃣ ranking


In [167]:
out = df.loc[~df['rank'].between(0, 5, inclusive='both'), ['brand','name','rank']]
print(f"rows out of [0.0,5.0]: {len(out)}")

rows out of [0.0,5.0]: 0


# Normalize 'Label'

In [168]:
df.columns = [col.lower() for col in df.columns]

# 2️⃣ แปลงค่าภายในคอลัมน์ 'label' ให้เป็นตัวพิมพ์เล็กทั้งหมด
if 'label' in df.columns:
    df['label'] = df['label'].astype(str).str.lower()

# Merge 'ingredients' column

In [169]:
def normalize_ing_string(s: str) -> str | pd._libs.missing.NAType:
    """
    รับสตริงสูตร → คืนสตริงที่สะอาด:
      - ตัดช่องว่างเกิน
      - split ด้วย ',', strip แต่ละตัว
      - ลบว่างเปล่า
      - ลบคำซ้ำโดยคงลำดับเดิม
      - join กลับด้วย ', '
    """
    if pd.isna(s):
        return pd.NA
    s = str(s)
    # normalize space
    s = re.sub(r'\s+', ' ', s).strip()
    # tokenize
    parts = [p.strip() for p in s.split(',')]
    parts = [p for p in parts if p]  # ตัดว่าง
    # ลบซ้ำแบบคงลำดับ
    seen = set()
    dedup = []
    for p in parts:
        if p not in seen:
            seen.add(p)
            dedup.append(p)
    return ', '.join(dedup) if dedup else pd.NA

def list_to_ing_string(x) -> str | pd._libs.missing.NAType:
    """
    รับ list ของ ingredients → คืนสตริงที่สะอาด (dedup + join)
    """
    if isinstance(x, list):
        # แปลงเป็นสตริงก่อนแล้ว normalize ต่อ
        return normalize_ing_string(', '.join(map(str, x)))
    # ถ้าไม่ใช่ list (เช่น เป็น NaN หรือเป็นสตริงอยู่แล้ว) ก็ normalize แบบสตริง
    return normalize_ing_string(x)

# 1) เตรียม candidate จาก ingredients_list (ถ้ามี)
if 'ingredients_list' in df.columns:
    cand_from_list = df['ingredients_list'].apply(list_to_ing_string)
else:
    cand_from_list = pd.Series(pd.NA, index=df.index)

# 2) เตรียม candidate จาก ingredients_cleaned (ถ้ามี)
if 'ingredients_cleaned' in df.columns:
    cand_from_cleaned = df['ingredients_cleaned'].apply(normalize_ing_string)
else:
    cand_from_cleaned = pd.Series(pd.NA, index=df.index)

# 3) รวมเป็นคอลัมน์แม่ 'ingredients'
#    เลือกค่าจาก ingredients_list ก่อน (เพราะลบซ้ำแล้ว)
#    ถ้าว่างค่อย fallback ไปใช้ ingredients_cleaned
ingredients_merged = cand_from_list.fillna(cand_from_cleaned)

# (ออปชัน) ถ้าคุณต้องการ “คงข้อมูลแถวที่น่าสงสัยไว้” ก็เซ็ตตรง ๆ ได้เลย:
df['ingredients'] = ingredients_merged

# (ออปชัน) ถ้าคุณต้องการ “ทิ้งแถวที่น่าสงสัย” ออกไปเลย:
# if 'invalid_ing' in df.columns:
#     df = df[~df['invalid_ing']].copy()         # เก็บเฉพาะแถวปกติ
#     df['ingredients'] = ingredients_merged.loc[df.index]

# 4) (ออปชัน) ลบคอลัมน์ช่วยที่ไม่ต้องการเก็บ
cols_to_drop = [
    'has_dup_ing', 'dup_list', 'ingredients_list',
    'dup_after_clean', 'ingredients_cleaned',
    'invalid_reason', 'invalid_ing',
    'rank_over_5'  # คุณบอกว่าแค่เช็ก ไม่จำเป็นต้องเก็บก็ลบได้
]
df = df.drop(columns=[c for c in cols_to_drop if c in df.columns], errors='ignore')


In [170]:
# แสดงเฉพาะคอลัมน์ที่เกี่ยวข้อง เพื่อดูว่ารวมกันถูกไหม
check_cols = ['ingredients_list', 'ingredients_cleaned', 'ingredients']
# ตรวจว่ามีคอลัมน์ไหนอยู่บ้างใน DataFrame
check_cols = [c for c in check_cols if c in df.columns]

# แสดงตัวอย่าง 10 แถวแรกแบบ side-by-side
print("🔍 ตัวอย่างผลการ merge ingredients")
display(df[check_cols].head(10))


🔍 ตัวอย่างผลการ merge ingredients


ingredients
0   algae (seaweed) extract | mineral oil | petrol...
1   galactomyces ferment filtrate (pitera) | butyl...
2   water | dicaprylyl carbonate | glycerin | cete...
3   algae (seaweed) extract | cyclopentasiloxane |...
4   water | snail secretion filtrate | phenyl trim...
5   water | saccharomyces/camellia sinensis leaf/c...
6   water | glycerin | caprylic/ capric triglyceri...
7   water | glycerin | cyclohexasiloxane | squalan...
10  caprylic/capric triglyceride dicaprylyl carbon...
11  water | dipropylene glycol | glycerin | methl ...

In [171]:
# แสดงทุกคอลัมน์ (ไม่ต้องเลือกเฉพาะบางคอลัมน์)
pd.set_option('display.max_columns', None)   # ให้ pandas แสดงทุกคอลัมน์
pd.set_option('display.width', 0)            # ป้องกันการตัดบรรทัด

# แสดงตัวอย่าง 10 แถวแรก
print("📋 ตัวอย่างข้อมูล 10 แถวแรกหลัง merge แล้ว:")
display(df.head(10))


📋 ตัวอย่างข้อมูล 10 แถวแรกหลัง merge แล้ว:


label              brand  \
0   moisturizer             la mer   
1   moisturizer               skii   
2   moisturizer     drunk elephant   
3   moisturizer             la mer   
4   moisturizer       it cosmetics   
5   moisturizer             tatcha   
6   moisturizer     drunk elephant   
7   moisturizer  kiehls since 1851   
10  moisturizer  kiehls since 1851   
11  moisturizer              belif   

                                           name   price  rank  \
0                               crème de la mer  6387.5   4.1   
1                      facial treatment essence  6533.5   4.1   
2                     protini polypeptide cream  2482.0   4.4   
3                   the moisturizing soft cream  6387.5   3.8   
4   your skin but better cc+ cream with spf 50+  1387.0   4.1   
5                               the water cream  2482.0   4.2   
6                      lala retro whipped cream  2190.0   4.2   
7                            ultra facial cream  1058.5   4.4   
10                midnight recovery concentrate  1715.5   4.4   
11                     the true cream aqua bomb  1387.0   4.5   

                                          ingredients  combination  dry  \
0   algae (seaweed) extract | mineral oil | petrol...            1    1   
1   galactomyces ferment filtrate (pitera) | butyl...            1    1   
2   water | dicaprylyl carbonate | glycerin | cete...            1    1   
3   algae (seaweed) extract | cyclopentasiloxane |...            1    1   
4   water | snail secretion filtrate | phenyl trim...            1    1   
5   water | saccharomyces/camellia sinensis leaf/c...            1    0   
6   water | glycerin | caprylic/ capric triglyceri...            1    1   
7   water | glycerin | cyclohexasiloxane | squalan...            1    1   
10  caprylic/capric triglyceride dicaprylyl carbon...            1    1   
11  water | dipropylene glycol | glycerin | methl ...            1    0   

    normal  oily  sensitive  
0        1     1          1  
1        1     1          1  
2        1     1          0  
3        1     1          1  
4        1     1          1  
5        1     1          1  
6        1     1          0  
7        1     1          1  
10       1     1          1  
11       1     1          0

In [172]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 811 entries, 0 to 1119
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   label        811 non-null    object 
 1   brand        811 non-null    object 
 2   name         811 non-null    object 
 3   price        811 non-null    float64
 4   rank         811 non-null    float64
 5   ingredients  811 non-null    object 
 6   combination  811 non-null    int64  
 7   dry          811 non-null    int64  
 8   normal       811 non-null    int64  
 9   oily         811 non-null    int64  
 10  sensitive    811 non-null    int64  
dtypes: float64(2), int64(5), object(4)
memory usage: 76.0+ KB


In [173]:
df.head(20)

label              brand  \
0   moisturizer             la mer   
1   moisturizer               skii   
2   moisturizer     drunk elephant   
3   moisturizer             la mer   
4   moisturizer       it cosmetics   
5   moisturizer             tatcha   
6   moisturizer     drunk elephant   
7   moisturizer  kiehls since 1851   
10  moisturizer  kiehls since 1851   
11  moisturizer              belif   
12  moisturizer       sunday riley   
13  moisturizer            farmacy   
14  moisturizer   first aid beauty   
15  moisturizer           clinique   
17  moisturizer               skii   
18  moisturizer             la mer   
21  moisturizer              fresh   
22  moisturizer              belif   
24  moisturizer            origins   
25  moisturizer           clinique   

                                                 name   price  rank  \
0                                     crème de la mer  6387.5   4.1   
1                            facial treatment essence  6533.5   4.1   
2                           protini polypeptide cream  2482.0   4.4   
3                         the moisturizing soft cream  6387.5   3.8   
4         your skin but better cc+ cream with spf 50+  1387.0   4.1   
5                                     the water cream  2482.0   4.2   
6                            lala retro whipped cream  2190.0   4.2   
7                                  ultra facial cream  1058.5   4.4   
10                      midnight recovery concentrate  1715.5   4.4   
11                           the true cream aqua bomb  1387.0   4.5   
12                            luna sleeping night oil  3832.5   4.1   
13  honeymoon glow aha resurfacing night serum wit...  2117.0   4.6   
14               ultra repair cream intense hydration  1095.0   4.6   
15  moisture surge 72-hour auto-replenishing hydrator  1423.5   4.4   
17                            r.n.a. power face cream  8395.0   4.3   
18                               crème de la mer mini  3102.5   4.1   
21                   black tea firming overnight mask  3358.0   4.1   
22                   the true cream moisturizing bomb  1387.0   4.6   
24  dr. andrew weil for origins mega-mushroom reli...  1241.0   4.4   
25        dramatically different moisturizing lotion+  1022.0   3.9   

                                          ingredients  combination  dry  \
0   algae (seaweed) extract | mineral oil | petrol...            1    1   
1   galactomyces ferment filtrate (pitera) | butyl...            1    1   
2   water | dicaprylyl carbonate | glycerin | cete...            1    1   
3   algae (seaweed) extract | cyclopentasiloxane |...            1    1   
4   water | snail secretion filtrate | phenyl trim...            1    1   
5   water | saccharomyces/camellia sinensis leaf/c...            1    0   
6   water | glycerin | caprylic/ capric triglyceri...            1    1   
7   water | glycerin | cyclohexasiloxane | squalan...            1    1   
10  caprylic/capric triglyceride dicaprylyl carbon...            1    1   
11  water | dipropylene glycol | glycerin | methl ...            1    0   
12  persea gratissima (extra virgin | cold pressed...            1    1   
13  water | lactic acid | propanediol | jojoba est...            1    1   
14  water | stearic acid | glycerin | c12-15 alkyl...            1    1   
15  water | dimethicone | butylene glycol | glycer...            1    1   
17  water | glycerin | galactomyces ferment filtra...            0    1   
18  algae (seaweed) extract | mineral oil | petrol...            1    1   
21  water | glycerin | butylene glycol | jojoba es...            1    1   
22  water | glycerin | cyclohexasiloxane | hydroge...            0    1   
24  water | butylene glycol | peg-4 | citrus auran...            1    1   
25  water | mineral oil/paraffinum liquidum/huile ...            1    1   

    normal  oily  sensitive  
0        1     1          1  
1        1     1          1  
2        1     1          0  
3        1     1       

In [174]:
df.shape

(811, 11)

In [177]:
df.to_csv("product_catalog_normalized.csv", index=False, encoding="utf-8-sig")

In [176]:
df.to_csv('product_catalog_clean.csv', index=False)